### PR TITLE
const_eval: Predetermine the layout of all locals when pushing a stack frame

### PR DESCRIPTION
**Pull request by dotdash**
Usually the layout of any locals is required at least three times, once
when it becomes live, once when it is written to, and once it is read
from. By adding a cache for them, we can reduce the number of layout
queries speeding up code that is heavy on const_eval.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Thanks for the PR, @dotdash! Looking forward to the perf results `:)`

**Comment by michaelwoerister** [comment]
This looks pretty good. For the synthetic const-eval heavy tests it looks spectacular even `:)`

I'll give it a proper review tomorrow. Meanwhile, could you check whether there's a problem with compiling the unicode_normalization crate? perf.rlo shows no numbers for it which might indicate a compilation error.

cc @oli-obk @rust-lang/wg-compiler-performance

**Comment by dotdash** [comment]
The numbers are only missing for the parent commit. I suppose that it just predates the addition of that benchmark, which was yesterday.

File: src/librustc_mir/interpret/eval_context.rs [hunk 1]
```diff
@@ -289,10 +290,8 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tc
         &self,
         frame: &Frame<'mir, 'tcx, M::PointerTag, M::FrameExtra>,
         local: mir::Local
-    ) -> EvalResult<'tcx, TyLayout<'tcx>> {
-        let local_ty = frame.mir.local_decls[local].ty;
-        let local_ty = self.monomorphize(local_ty, frame.instance.substs);
-        self.layout_of(local_ty)
+    ) -> TyLayout<'tcx> {
+        frame.local_layouts[local]
```
Comment (oli-obk) [inline src/librustc_mir/interpret/eval_context.rs]: While definitely an improvement for all currently possible const eval code, I think this is suboptimal for code with branching, as a lot of locals will never be touched during evaluation. I think it would be better to make `local_layouts` be `IndexVec<mir::Local, Option<TyLayout<'tcx>>>` and only fill it in on demand. To make this really work with reading I presume it would even need to be `IndexVec<mir::Local, RefCell<Option<TyLayout<'tcx>>>>` at which point we can just move to `IndexVec<mir::Local, Once<TyLayout<'tcx>>>` or some non-Sync variant of that.

File: src/librustc_mir/interpret/operand.rs [hunk 2]
```diff
@@ -463,12 +463,10 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
         &self,
```
Comment (oli-obk) [inline src/librustc_mir/interpret/operand.rs]: drive-by fix: `access_local` doesn't need to be `pub` anymore

**Comment by michaelwoerister** [comment]
@oli-obk's remark regarding branching code makes sense. Would it be possible to make some or most `EvalContext` methods `&mut self` instead of `&self` so we don't have to use `RefCell`? It seems like that should be possible in principle, right?

**Comment by oli-obk** [comment]
> Would it be possible to make some or most EvalContext methods &mut self instead of &self so we don't have to use RefCell? It seems like that should be possible in principle, right?

The problem is that we separated reading and writing into `&mut self` and `&self` methods. So any method trying to read from the local would need to be `&mut self` and that would infect everything. The immutability of the read paths has been very helpful in the development of miri. I think it would be fine to have some interior mutability just for memoization. Since `TyLayout` is `Copy`, we can even use `Cell<Option<TyLayout<'tcx>>>` making the entire operation zero cost both in memory and performance.

**Comment by michaelwoerister** [comment]
OK, I'm fine with the `Cell` approach.

I'd be interested though in how using `&self` has made things easier.

**Comment by oli-obk** [comment]
> I'd be interested though in how using &self has made things easier.

Mirroring the mutability of the virtual miri memory in the mutability of the memory datastructure itself allows various things like holding references to different parts of the virtual memory at the same time. If we made everything take mutable references, we could only ever access one miri allocation, even if we just wanted to write to one. While we could have worked around this with unsafe code (and we still have to do it in one situation https://github.com/rust-lang/rust/blob/38650b69cafbff61d71a275ced1e9866a08a36c0/src/librustc_mir/interpret/memory.rs#L725-L763), and while two phase borrows could have made many situations safe, we would still have to fight the borrow checker at ever turn.

Also it just "felt right" to make sure we don't accidentally modify our virtual memory in functions that should just read from it.

**Comment by michaelwoerister** [comment]
Yeah, I can see how this can become tricky. Thanks for the explanation, @oli-obk!

**Comment by dotdash** [comment]
Seems that the constraint on the stable hash impl was just swapped around, asking for 'mir to outlive 'tcx instead of the other way around.

**Comment by dotdash** [comment]
@bors try

We should measure this again, my local setup seems a bit funky right now, and showed a massive slowdown for this version which seems unlikely to be right.

**Comment by oli-obk** [comment]
Perf is great (though no effect on unicode-normalization). Impl looks good.

The commit message is still slightly outdated.

**Comment by michaelwoerister** [comment]
@bors r+

Thanks, @dotdash!

File: src/librustc_mir/interpret/operand.rs [hunk 3]
```diff
@@ -510,7 +504,7 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
             // FIXME: do some more logic on `move` to invalidate the old location
             Copy(ref place) |
             Move(ref place) =>
-                self.eval_place_to_op(place, layout)?,
+                self.eval_place_to_op(place)?,
```
Comment (RalfJung) [inline src/librustc_mir/interpret/operand.rs]: Hm, but this means we might have to compute the layout even if we already know it, right?

Comment (oli-obk) [reply src/librustc_mir/interpret/operand.rs]: To change this we'd need to pass a precomputed layout to `layout_of_local`, which probably doesn't do anything, because any initialized local will already have had called `layout_of_local` on it before. We can still benchmark it and see if it changes anything

File: src/librustc_mir/const_eval.rs [hunk 4]
```diff
@@ -72,6 +72,7 @@ fn mk_eval_cx_inner<'a, 'mir, 'tcx>(
     ecx.stack.push(interpret::Frame {
         block: mir::START_BLOCK,
         locals: IndexVec::new(),
+        local_layouts: IndexVec::new(),
```
Comment (eddyb) [inline src/librustc_mir/const_eval.rs]: Can't the information be included in `locals` itself?
